### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by richfitz**
~Merge after #6, as this PR builds on that one.~

I've had to slightly rejig how the `base` object was created because otherwise there were stray `__module` bits within it. This way seems reasonable though?

There are more exports than before - that's because typedoc tells you when you've referenced something that is not exported - most of these are just types.

Otherwise this PR should be minor formatting changes plus addition of doc comments

Docs online here: https://mrc-ide.github.io/odin-js/

### CODE REVIEW

File: src/maths.ts [hunk 1]
```diff
@@ -6,6 +6,29 @@ function roundHalfToEven(x: number) {
     }
 }
 
+/** Round a number. This function diffes from `Math.round` in two
```
Comment (EmmaLRussell) [inline src/maths.ts]:
```suggestion
/** Round a number. This function differs from `Math.round` in two
```

File: src/model.ts [hunk 2]
```diff
@@ -5,30 +5,147 @@ import {base, BaseType} from "./base";
...
-interface OdinModelODE {
-    setUser(pars: UserType, unknownAction: string): void;
+    new(base: BaseType, pars: UserType, unusedUserAction: string) => OdinModel;
+
+/**
+ * The interface that odin ordinary differential equation (ODE) models
+ * will match, though they will be generated in plain JavaScript and
+ * not TypeScript.
+ */
+export interface OdinModelODE {
```
Comment (EmmaLRussell) [inline src/model.ts]: Are OdinModelODE and OdinModelDDE actually identical (or nearly?) - could this just be one interface, or extended from a common interface? Should have asked this in the last PR really...

File: src/model.ts [hunk 3]
```diff
@@ -5,30 +5,147 @@ import {base, BaseType} from "./base";
...
+    /**  Set parameters into an existing model.
+     *
+     * @param pars New parameters for the model. Values are are
+     * ommitted here but present in the model will be unchanged.
```
Comment (EmmaLRussell) [inline src/model.ts]:
```suggestion
     * omitted here but present in the model will be unchanged.
```

File: src/model.ts [hunk 4]
```diff
@@ -5,30 +5,147 @@ import {base, BaseType} from "./base";
...
+/**
+ * The interface that odin delay differential equation (DDE) models
+ * will match, though they will be generated in plain JavaScript and
+ * not TypeScript.
+ */
+export interface OdinModelDDE {
+    /**  Set parameters into an existing model.
+     *
+     * @param pars New parameters for the model. Values are are
+     * ommitted here but present in the model will be unchanged.
```
Comment (EmmaLRussell) [inline src/model.ts]:
```suggestion
     * omitted here but present in the model will be unchanged.
```

File: src/model.ts [hunk 5]
```diff
@@ -5,30 +5,147 @@ import {base, BaseType} from "./base";
...
+     *
+     * @param unusedUserAction String, describing the action to take
+     * if there are unknown values in `pars` - possible values are
+     * "error", "ignore", "warning" and "message"
```
Comment (EmmaLRussell) [inline src/model.ts]: Could make that a string enum I guess

File: src/user.ts [hunk 6]
```diff
@@ -31,11 +82,34 @@ export function checkUser(user: UserType, allowed: string[],
...
+ * @param name Name of the parameter to set
+ *
+ * @param internal The model's internal data; the parameter will be
+ * updated here
+ *
+ * @param default A default value for the parameter, or `null` if the
+ * parameter is required
+ *
+ * @param min The minimum allowed value for the parameter; use
+ * `-Infinity` there is no minimum
```
Comment (EmmaLRussell) [inline src/user.ts]:
```suggestion
 * `-Infinity` if there is no minimum
```

File: src/user.ts [hunk 7]
```diff
@@ -31,11 +82,34 @@ export function checkUser(user: UserType, allowed: string[],
...
+ *
+ * @param max The maximum allowed value for the parameter; use
+ * `Infinity` there is no maximum
```
Comment (EmmaLRussell) [inline src/user.ts]:
```suggestion
 * `Infinity` if there is no maximum
```

File: src/user.ts [hunk 8]
```diff
@@ -31,11 +82,34 @@ export function checkUser(user: UserType, allowed: string[],
...
+ *
+ * @param isInteger Check that the provided value is a parameter
```
Comment (EmmaLRussell) [inline src/user.ts]:
```suggestion
 * @param isInteger Check that the provided value is an integer```

File: src/user.ts [hunk 9]
```diff
@@ -53,13 +127,36 @@ export function setUserScalar(user: UserType, name: string,
...
+ * the remaining values being the length of each dimension. The first
+ * number will therefore be the product of the remaining numbers.
+ *
+ * @param min The minimum allowed value for the parameter; use
+ * `-Infinity` there is no minimum
+ *
+ * @param max The maximum allowed value for the parameter; use
+ * `Infinity` there is no maximum
+ *
+ * @param isInteger Check that the provided value is a parameter
```
Comment (EmmaLRussell) [inline src/user.ts]:
```suggestion
 * @param isInteger Check that the provided value is an integer
```

File: src/model.ts (hunk 2 above)
Comment (richfitz) [reply src/model.ts]: They differ in their rhs and output methods. We could use a base method and extend both (probably in a new PR though?)

File: src/model.ts (hunk 5 above)
Comment (richfitz) [reply src/model.ts]: yeah, I noted this down but had not made a ticket for it. Now up at https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-3357

File: src/user.ts (hunk 8 above)
Comment (richfitz) [reply src/user.ts]: :facepalm:

File: src/model.ts (hunk 2 above)
Comment (richfitz) [reply src/model.ts]: actually this is not bad to implement, will do it now

Comment (richfitz) [reply src/model.ts]: Done in [8922fb1](https://github.com/mrc-ide/odin-js/pull/7/commits/8922fb1508d4dd7ac84cb728890f3c4af5dcef4f)

File: src/user.ts [hunk 10]
```diff
@@ -98,12 +98,12 @@ export function checkUser(pars: UserType, allowed: string[],
...
  * @param min The minimum allowed value for the parameter; use
- * `-Infinity` there is no minimum
+ * `-Infinity` if there is no minimum
  *
  * @param max The maximum allowed value for the parameter; use
- * `Infinity` there is no maximum
+ * `Infinity` if there is no maximum
  *
- * @param isInteger Check that the provided value is a parameter
+ * @param isInteger Check that the provided value is an integer```
```
Comment (EmmaLRussell) [inline src/user.ts]: rogue backticks here - think they were probably my fault!

**Review by EmmaLRussell — APPROVED** [review]

**Pull request merged**